tokio-util/src/sync/cancellation_token.rs

```diff
@@ -319,560 +190,374 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
+    ///
+    /// Sadly, it's impossible to figure out the parent of a node without
+    /// locking it. To then achieve locking order consistency, the node
+    /// has to be unlocked before the parent gets locked.
+    /// This leaves a small window where we already assume that we know the parent,
+    /// but neither the parent nor the node is locked. Therefore, the parent could change.
+    ///
+    /// To prevent that this problem leaks into the rest of the code, it is abstracted
+    /// in this function.
+    ///
+    /// The locked child and optionally its locked parent, if a parent exists, get passed
+    /// to the `func` argument via (node, None) or (node, Some(parent)).
+    fn with_locked_node_and_parent<F, Ret>(node: &Arc<TreeNode>, func: F) -> Ret
     where
-        F: Fn(StateSnapshot) -> StateSnapshot,
+        F: FnOnce(MutexGuard<'_, Inner>, Option<MutexGuard<'_, Inner>>) -> Ret,
     {
-        let mut current_packed_state = current_state.pack();
-        loop {
-            let next_state = func(current_state);
-            match self.state.compare_exchange(
-                current_packed_state,
-                next_state.pack(),
-                Ordering::SeqCst,
-                Ordering::SeqCst,
-            ) {
-                Ok(_) => {
-                    return next_state;
-                }
-                Err(actual) => {
-                    current_packed_state = actual;
-                    current_state = StateSnapshot::unpack(actual);
-                }
+        let mut potential_parent = {
+            let locked_node = node.inner.lock().unwrap();
+            match locked_node.parent.clone() {
+                Some(parent) => parent,
+                // If we locked the node and its parent is `None`, we are in a valid state
+                // and can return.
+                None => return func(locked_node, None),
             }
-        }
-    }
+        };
 
-    fn increment_refcount(&self, current_state: StateSnapshot) -> StateSnapshot {
-        self.atomic_update_state(current_state, |mut state: StateSnapshot| {
-            if state.refcount >= MAX_REFS as usize {
-                eprintln!("[ERROR] Maximum reference count for CancellationToken was exceeded");
-                std::process::abort();
+        loop {
+            // Deadlock safety:
+            // At this very point, we *assume* that 'potential_parent' is the parent of 'node'.
+            // But both are currently unlocked. Therefore, we cannot rely any more that this is the case.
+            // Even worse, 'potential_parent' might actually become a *sibling* of 'node', which would violate
+            // our 2. rule of deadlock safety.
+            //
+            // However, we can prove that the lock order has to be 'potential_parent' first, then 'node',
+            // because based on invariant #2, 'node' is younger than 'potential_parent', as
+            // 'potential_parent' once was the parent of 'node'.
```

Comment:
I'm not a fan of the way this is phrased. You are saying "this doesn't work because X, but actually it works anyway because Y". I would prefer something like "this works because Y. It might seem like X is an issue, but reasons"

Reply:
Makes sense. Will rephrase.

Reply:
Fixed, potentially